packages/conferer/src/Conferer/Config/Internal/Types.hs

```diff
@@ -26,13 +32,68 @@ data Config =
   } deriving (Show)
 
 -- | Result of a key lookup in a 'Config'
-data KeyLookupResult
-  = MissingKey [Key]
-  | FoundInSources Key Text
-  | FoundInDefaults Key [Dynamic]
-  deriving (Show)
+--
+-- This is indexed by 'LookupTarget' to indicate where we are looking for values
+-- and some invariants (if you look for defaults you'll never get things from a
+-- 'Source)
+--
+-- Possible valures are:
+--   * 'BothSourcesAndDefaults' a: All constructors are possible, type parameter
+--       indicates the type of the default to look for.
+--   * 'OnlyDefaultsAs' a: 'FoundInSources' constructor is not possible, type
+--       parameter indicates the type of the default to look for
+--   * 'OnlySources': 'FoundInDefaults' constructor is not possible, type of
+--       of the result is always 'Text'
+data KeyLookupResult lookupTarget
+  = MissingKey (MissingC lookupTarget) [Key]
+  | FoundInSources (SourcesC lookupTarget) Key
+  | FoundInDefaults (DefaultsC lookupTarget) Key
+
+-- | Target of a lookup that the 'KeyLookupResult' is parameterized on
+-- to indicate where the value was looked for in.
+data LookupTarget a
+ = BothSourcesAndDefaults a
+ -- ^ Look in both Sources and defaults (this is the most common)
+ -- its type parameter indicates the type of the default
+ | OnlySources
+ -- ^ Look only in the defaults ('FoundInDefaults' is not possible)
+ | OnlyDefaultsAs a
+ -- ^ Look only in the sources ('FoundInSources' is not possible)
+ -- its type parameter indicates the type of the default
+
+-- | Modifier type family for the 'MissingKey' constructor.
+type family MissingC (a :: LookupTarget Type) where
+  MissingC ('BothSourcesAndDefaults a) = ()
+  MissingC 'OnlySources = ()
+  MissingC ('OnlyDefaultsAs _) = ()
+
+-- | Modifier type family for the 'FoundInSources' constructor.
+type family SourcesC (a :: LookupTarget Type) where
+  SourcesC ('BothSourcesAndDefaults a) = Text
+  SourcesC 'OnlySources = Text
+  SourcesC ('OnlyDefaultsAs _) = Void
+
+-- | Modifier type family for the 'FoundInDefaults' constructor.
+type family DefaultsC (a :: LookupTarget Type) where
+  DefaultsC ('BothSourcesAndDefaults a) = a
+  DefaultsC 'OnlySources = Void
+  DefaultsC ('OnlyDefaultsAs a) = a
```

Comment:
what did the C mean in `MissingC`, `SourcesC` and `DefaultsC`?

I feel like this seems to represent the kind of result in the search, like:
`MissingKey`, `FoundInSources` and `FoundInDefaults`. What do you think?

Ohh, I just saw that they always go along with constructor of those exact names 😅 , so it actually represents that but on a different level. I think it´d be better to have these named with similar names to the constructors of `KeyLookupResult`

Reply:
That's exactly right, MissingC is Missing constructor, and yes, it's a pretty bad name, I'll try to think of something better

Reply:
So I've got something better, I want to follow that those semantics that you said: type family is the result type of the lookup.

With that in mind:

* I've removed the `MissingC` type family since it doesn't fit those semantics (if the value is missing the value will never be present) but I've left the `()` in the constructor in case we want to extend it in the future with less breakage)
* I've renamed `SourceC` and `DefaultsC` to `SourcesResultType` and `DefaultsResultType` respectively

Not sure if this is that much better but it's a start